File: app/src/main/java/com/nacoda/moviesmvvm/data/source/MoviesDataSource.java
package com.nacoda.moviesmvvm.data.source;

import com.nacoda.moviesmvvm.data.model.Casts;
import com.nacoda.moviesmvvm.data.model.Detail;
import com.nacoda.moviesmvvm.data.model.Movie;

import java.util.List;

/**
 * Created by Mayburger on 1/9/18.
 */

public interface MoviesDataSource {

        void getPopular(GetMoviesCallback callback);
        void getNowPlaying(GetMoviesCallback callback);
        void getTopRated(GetMoviesCallback callback);
        void getSearch(GetMoviesCallback callback, String query);
        void getDetail(GetDetailCallback callback, String movieId);
        void getCasts(GetCastsCallback callback, String movieId);


    interface GetMoviesCallback {
        void onMoviesLoaded(List<Movie> movies);
        void onDataNotAvailable();
        void onError(String errorMessage);
    }
    interface GetDetailCallback {
        void onDetailLoaded(Detail detail);
        void onDataNotAvailable();
        void onError(String errorMessage);
    }

    interface GetCastsCallback {
        void onCastsLoaded(Casts casts);
        void onDataNotAvailable();
        void onError(String errorMessage);
    }

}

File: app/src/main/java/com/nacoda/moviesmvvm/data/source/MoviesDataSource.kt
package com.nacoda.moviesmvvm.data.source

import com.nacoda.moviesmvvm.data.model.Casts
import com.nacoda.moviesmvvm.data.model.Detail
import com.nacoda.moviesmvvm.data.model.Movie


/**
 * Created by irfanirawansukirman on 04/12/17.
 */
interface MoviesDataSource {

    /**
     * Get all movie list
     * @param callback
     */
    fun getPopular(callback: GetMoviesCallback)

    fun getNowPlaying(callback: GetMoviesCallback)

    fun getTopRated(callback: GetMoviesCallback)

    fun getSearch(callback: GetMoviesCallback, query: String)

    fun getDetail(callback: GetDetailCallback, movieId: String)

    fun getCasts(callback: GetCastsCallback, movieId: String)

    interface GetDetailCallback {
        fun onDetailLoaded(detail: Detail?)
        fun onDataNotAvailable()
        fun onError(errorMessage: String?)
    }

    interface GetCastsCallback {
        fun onCastsLoaded(casts: Casts)
        fun onDataNotAvailable()
        fun onError(errorMessage: String?)
    }

    interface GetMoviesCallback {

        fun onMoviesLoaded(movies: List<Movie>?)

        fun onDataNotAvailable()

        fun onError(errorMessage: String?)
    }
}